### ISSUE TITLE
When and why are tiddler field-names forced to be lowercase?

### ISSUE THREAD

**Issue by tobibeer** [Jermolene/TiddlyWiki5#2155]
HTML5 appears to respect / allow both cases.

If field-names are indeed forced to be all-lowercase, then **tiddler.getFieldsString()** must always turn a requested field-name string to lowercase first, before checking.

**Context:** [discussion](https://groups.google.com/d/msg/tiddlywiki/6SyZtodCsm4/iIwffM7BBwAJ)

**Comment by felixhayashi** [comment]
> If field-names are indeed forced to be all-lowercase, then tiddler.getFieldsString() must always turn a requested field-name string to lowercase first, before checking.

Wooooo please don't. I actually create many temp tiddlers with camel case field names. TW caches the tiddler objects with camel case names and doesn't force the fields to be lowercase (only when saving and temp tiddlers are not saved!!).

> When and why are tiddler field-names forced to be lowercase?

Not sure either. I think it is for a cosmetic reason or maybe the conversion in a specific format during export depends on it?

-Felix

**Comment by Evolena** [comment]
It was in [5.0.0-alpha-16](http://tiddlywiki.com/#Release%205.0.0-alpha.16) that it was forced in the UI (more precisely in b8f15fe), but I don't know why.

**Comment by matabele** [comment]
Not everywhere!

```
<$button>
<$action-setfield $field="Donkey" $value="ass"/>
Set
</$button>
```
But, don't do this -- the text field gets wiped :-/

```
<$button>
<$action-setfield Donkey="ass"/>
Set
</$button>
```

**Comment by felixhayashi** [comment]
@Evolena 

> It was in 5.0.0-alpha-16 that it was forced in the UI (more precisely in b8f15fe)

Thanks for the info!

@matabele 

That is what I meant in my comment (https://github.com/Jermolene/TiddlyWiki5/issues/2155#issuecomment-167268699). During runtime the uppercase stays only when saved, lowercase is enforced. I use camelcase field names a lot in temp tiddler in TiddlyMap as it is very convenient. If there is not reason for forcing lowercase, I would even vote to abandon the rule as it is clearly allowed to have HTML attributes contain uppercase letters and I do not see why *.tid files should not be able to hold uppercase property names…

**Comment by matabele** [comment]
OK -- I submitted a PR to fix the inadvertent clearing of the 'text' field: https://github.com/Jermolene/TiddlyWiki5/pull/2165

**Comment by tobibeer** [comment]
> OK -- I submitted a PR to fix the inadvertent clearing of the 'text' field: #2165

I think this is "fixing" the wrong problem, though... and potentially should not be changed as it may actually be used as a feature, rather than a bug.

**Comment by matabele** [comment]
@tobibeer 

In what circumstance would you wish to clear the 'text' field of the tiddler from which the ActionSetfield widget is called? 

The problem arise when you use the short form of SetText only:

```
<$button>
<$action-setfield field1="something" field2="something else"/>
Set Some Fields
</$button>
```
The condition could be made more restrictive -- to include only the case when neither the `$tiddler` nor the `$field` attribute values are set.

**Comment by tobibeer** [comment]
> In what circumstance would you wish to clear the 'text' field of the tiddler from which the ActionSetfield widget is called?

For example, in combination with a list or tiddler widget providing a button with respect to some tiddler. Just because you haven't considered it, doesn't necessarily mean nobody would use it this way... as this is quite how it works, atm. If something works as defined, and someone makes use of it... why wouldn't they? The docs for action-setfield clearly state what the defaults are. So, all those put together, the effect is pretty clear. What isn't clear, though, is that we're not allowed uppercase field-names.

**Comment by matabele** [comment]
@tobibeer 

The only purpose for this would be to clear the text fields of every tiddler listed with the $list widget -- and in that case, it would be safer to have to specify `field="text"`. My fix only checks for the condition that the`field`attribute has been set before clearing a text field.

**Comment by felixhayashi** [comment]
@matabele 

> OK -- I submitted a PR to fix the inadvertent clearing of the 'text' field: #2165

I am a bit baffled. This issue is about the question why field-names are not allowed to be uppercase. How is that related to the `SetFieldWidget`? This widget can set the field, yes, but I can also create a tiddler with uppercase field names and add it using `addTiddler()`, when saving and refreshing, the js-runtime-tiddler-store of the last session is cleared and fields are lowercase.

**Comment by pmario** [comment]
Here's the reason why field names are lowercase only and have some other restrictions. It's forced by the implementation at the moment:  https://www.youtube.com/watch?v=FDV5ojcCLB8#

It's possible to fix the problem, but it would need a redesign of, how TW saves tiddler content. Jeremy shows the reasons in the video. So don't expect it in the near future.

**Comment by matabele** [comment]
@felixhayashi 

The fix is nothing to do with lowercase fieldnames -- I was pointing out that a field name may be created with mixed case with the ActionSetField widget, but reverts to lowercase when saved and refreshed. 

Whilst experimenting I noticed that my code would get wiped when I used the short form SetText. The fix was by the way.

**Comment by felixhayashi** [comment]
> Here's the reason why field names are lowercase only and have some other restrictions. It's forced by the implementation at the moment: https://www.youtube.com/watch?v=FDV5ojcCLB8#

Many thanks @pmario for pointing to this resource (and again it shows how useful video annotations are!).

In the video, @Jermolene clearly says in response to Eric's question that the restriction is due to the fact that field names are stored as HTML attributes (See minute 23:40 of the video). Additionally the colon (`:`) is enforced due to the saving in `*.tid` format. **So apart from this uppercase is clearly valid and I still see no point in enforcing lowercase.**

**So in summary:** The only restrictions should be non-valid HTML attribute names **AND** colons.

-Felix

**Comment by felixhayashi** [comment]
> The fix is nothing to do with lowercase fieldnames […] Whilst experimenting I noticed that my code would get wiped when I used the short form SetText. The fix was by the way.

Ah, k, I get it.

**Comment by pmario** [comment]
>So in summary: The only restrictions should be non-valid HTML attribute names AND colons. – Uppercase/Camelcase should be allowed.

Nope! see: http://www.w3.org/TR/html-markup/syntax.html#syntax-attributes

> There must never be two or more attributes on the same start tag whose names are a case-insensitive match for each other.

Uppercase and CamelCase violate this _must_ rule, since `this`, `This` and `ThIs` would be valid atributes for one tiddler, but are not valid HTML!

**Comment by tobibeer** [comment]
Just to reiterate, since no two attributes should match when compared case-insensitively, I am fine with procedurally storing them in all-lowercase.

So, the to make this approach wholesome, all code-bits that are to retrieve and match field-names must first convert a given field-name to be queried to lowercase.

It should be permissible for a user to write things in uppercase, whether in the editor or in wikitext ...without any compromised behavior caused by that.

On the other hand, it is quite clear now that not even with CamelCased titles only you can have any kind of tiddler-name &lt;&gt; field-name mapping.

**Comment by felixhayashi** [comment]
@pmario 

> There must never be two or more attributes on the same start tag whose names are a case-insensitive match for each other.

Wow, good catch!

> So, to make this approach slightly more wholesome, all code-bits that are to retrieve and match field-names must first convert a given field-name to be queried to lowercase.

I would suggest to leave everything as is now, i.e.

* the editor enforces lowercase
* TW enforces lowercase when the wiki is saved.

**Comment by tobibeer** [comment]
It's slightly confusing that TiddlyWiki would sanitize things in some cases while it would break in others. Ideally, it should not break in any case... when it comes to case sensitivity over field-names. @Jermolene, if you think it's worth making sure this doesn't happen, I'll look into it, otherwise I'll close this issue for now.

**Comment by pmario** [comment]
uups. as Felix found out the discussion about field names start here:  https://youtu.be/FDV5ojcCLB8?t=1253 ... I fixed the link in my first post for future reference. Sorry. I didn't test it :(
